297 arch overview (#24)
Kuadrant/architecture: docs/design/architectural-overview.md

* The data plane is where kuadrant provided policy enforcement points are installed and integrated with a running gateway. 
* Other policies such as rate limit policy and auth policy are reconciled in the control plane and converted into configuration to be consumed by the data plane components.

## 1000m Architecture

![](./images/1000m-overview.png)

**Note:**
Currently some of our policies are focused only around the multi-cluster context (DNSPolicy and TLSPolicy in particular). It is high on our roadmap priority list to ensure our policies can operate at both the single and multi-cluster level.

### Control Plane Components and Responsibilities
A control plane component is something responsible for accepting instruction via a CRD based API and ensuring that configuration is manifested into state that can be acted on.

- **[Kuadrant Operator](https://github.com/Kuadrant/kuadrant-operator):**
    - Installation of data plane components via their respective operators
    - Exposes RateLimitPolicy and AuthPolicy and is currently the policy controller for these APIs
    - Configures the Gateway to be able to leverage the data plane enforcement components

> - **[Multi-Cluster Gateway Controller](https://github.com/Kuadrant/multicluster-gateway-controller):**

Author: not sure I follow? Multi-Cluster gateway controller is part of the current architecture?

> - **[Multi-Cluster Gateway Controller](https://github.com/Kuadrant/multicluster-gateway-controller):**

Reviewer: It is but there was a talk about unifying the operators in the future so I wanted to ask if this doc is supposed to cover these changes now or if it will be part of the future RFC

> - **[Multi-Cluster Gateway Controller](https://github.com/Kuadrant/multicluster-gateway-controller):**

Author: ah ok. I call that out under Future Direction at the end. I am fleshing that out further. The details of how that will happen will end up in an RFC though

- **[Limitador Operator:](https://github.com/Kuadrant/limitador-operator)**
    - Installs and configures Limitador
- **[Authorino Operator:](https://github.com/Kuadrant/authorino-operator)**
    - Installs and configures Authorino

### Data Plane Components and Responsibilities
A data plane component sits in the request flow and is responsible for enforcing policy configuration as a request comes through a gateway

- **Limitador:**
    - Complies with the with Envoy rate limiting API to provide rate limiting to the gateway
- **Authorino:**
    - Complies with the Envoy external auth API to provide auth integration to the gateway
- **WASM Plugin:**
    - Uses the Proxy WASM ABI Spec to integrate with Envoy and provide filtering and connectivity to authorino and limitador for request time enforcement of auth and rate limiting

### Dependencies and integrations

In order to provide its full suite of functionality, Kuadrant has several depenencies

- **[Cert-Manager](https://cert-manager.io/):** (provides the TLS integration)
    - required if you want to use TLSPolicy
- **[Open Cluster Manager](https://open-cluster-management.io/):** (provides a multi-cluster control plane)

>     - required if you want to use multi-cluster gateways

Author: Yes it is still needed currently even in single cluster to use DNS and TLS policy. Will make clearer

## Key Architectural Areas

* The Kuadrant architecture is spread across a control plane and also a data plane. Kuadrant can work in both a single and multi-cluster context. 
Currently in order for all APIs to work in a single or multi-cluster context you need to have OCM installed.  While this may change in the future, this approach allows us to start with a single cluster and seamlessly scale as more clusters are added.
* The control plane is where policies are exposed and expressed as kubernetes APIs and reconciled by the Kuadrant policy controllers. 
* The data plane is where Kuadrant's service protection components, configured by the  control plane policies, are enforced within the gateway instance as part of the request flow. 

> ## 1000m Architecture

Reviewer 2: what does 1000m Architecture mean?

Author: haha well exactly that. If you were to look at it from 1000M or basically a very high level :)

Reviewer 2: Ohhhhh ok i like that, that's cool

> # Kuadrant Architectural Overview [Draft]

Reviewer 3: Nowhere is described the effective (HHTP) flow of a request to a gateway configured by Kuadrant (showing the multi cluster aspect, rate limiting etc). That projection / view could be useful.

Author: hmm so perhaps this image is not clear enough or we need a separate one entirely https://github.com/Kuadrant/architecture/pull/24/files#diff-0997a2bfd6775f89fe2228aabc6093a9a547c746f1e778e5731da5e4563287e1

## Overview

It is important to note that Kuadrant is not in itself a gateway provider. Kuadrant provides a set of valuable policy APIs that enhance [Gateway API](https://github.com/kubernetes-sigs/gateway-api) via its defined [policy attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment/) extension point. The policy APIs are reconciled by a set of policy controllers and enforced via integration at different points to configure, enhance and secure the application connectivity provided via Gateway API and the underlying gateway provider. 

> These policy extensions are focused around areas such as DNS management supporting global load balancing and health checks, alongside service protection specific APIs such as rate limiting and auth. Kuadrant also integrates with [OCM (Open Cluster Manager)](https://open-cluster-management.io/) as a multi-cluster control plane to enable defining and distributing Gateways across multiple clusters, providing load balancing and tls management for these distributed gateways alongside the single view and single definition of gateways from the OCM management hub. 

Reviewer 4: How about something like this:
```
... providing load balancing and TLS management for these distributed gateways.

This integration can be managed centrally in a declarative way from the OCM Hub using kubernetes resources. 
```

## Key Architectural Areas

* The Kuadrant architecture is spread across a control plane and also a data plane. Kuadrant can work in both a single and multi-cluster context. 

> Currently in order for all APIs to work in a single or multi-cluster context you need to have OCM installed.  While this may change in the future, this approach allows us to start with a single cluster and seamlessly scale as more clusters are added.

Reviewer 3: You know, I know it's more to type but making explicit OCM to Open Cluster Manager would help readability I think.

* The control plane is where policies are exposed and expressed as kubernetes APIs and reconciled by the Kuadrant policy controllers. 

> * The data plane is where Kuadrant's service protection components, configured by the  control plane policies, are enforced within the gateway instance as part of the request flow. 

Reviewer 3: ```suggestion
* The data plane is where Kuadrant's service protection components, configured by the control plane policies, are enforced within the gateway instance as part of the request flow. 
```

## 1000m Architecture

![](./images/1000m-overview.png)

> **Note:**

Reviewer 3: What make you feel that such transient warning was to be put in the architecture document?

Author: Yes this is fair. Really this should be something reflected in a different document (IE some roadmap)

Currently some of our policies are focused only around the multi-cluster context ([DNSPolicy](https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/docs/dns-policy.md) and [TLSPolicy](https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/docs/tls-policy.md)). It is high on our road map priority list to ensure all of our policies can operate at both the single and multi-cluster level.

### Control Plane Components and Responsibilities

> A control plane component is something responsible for accepting instruction via a CRD based API and ensuring that configuration is manifested into state that can be acted on.

Reviewer 3: ```suggestion
Control plane components of Kuadrant are responsible for accepting instruction via a CRD based API and ensuring that configuration is manifested into state that can be acted on.
```

- **[Kuadrant Operator](https://github.com/Kuadrant/Kuadrant-operator):**
    - Installation of data plane service protection components via their respective operators
    - Exposes `RateLimitPolicy` and `AuthPolicy` and is currently the policy controller for these APIs
    - Configures the Gateway to be able to leverage the data plane service protection components
- **[Multi-Cluster Gateway Controller](https://github.com/Kuadrant/multicluster-gateway-controller):**
    - Exposes `DNSPolicy` and `TLSPolicy`
    - Configures DNS providers (e.g AWS Route 53) and TLS providers
    - Focused around use cases involving distributed gateways (for example across clouds or geographic regions)

>     - Integrates with OCM as the multi-cluster management hub to distribute and observe gateway status based on the clusters they are deployed to works directly with OCM APIs such `PlacementDecision` and `ManifestWork`.

Reviewer 3: ```suggestion
    - Integrates with OCM as the multi-cluster management hub to distribute and observe gateway status based on the clusters they are deployed to works directly with OCM APIs such as `PlacementDecision` and `ManifestWork`.
```

- **[Kuadrant-add-on-manager](https://github.com/Kuadrant/multicluster-gateway-controller/cmd/ocm):**
    - Sub component in the gateway controller repository
    - Follows the [add-on pattern](https://open-cluster-management.io/concepts/addon/) from OCM
    - Responsible for configuring and installing Kuadrant into a target spoke cluster

- **[Limitador Operator:](https://github.com/Kuadrant/limitador-operator)**
    - Installs and configures Limitador
- **[Authorino Operator:](https://github.com/Kuadrant/authorino-operator)**
    - Installs and configures Authorino

### Data Plane Components and Responsibilities
A data plane component sits in the request flow and is responsible for enforcing policy and providing service protection capabilities based on configuration managed and created by the control plane.

- **Limitador:**
    - Complies with the with Envoy rate limiting API to provide rate limiting to the gateway
- **Authorino:**
    - Complies with the Envoy external auth API to provide auth integration to the gateway
- **WASM Plugin:**

>     - Uses the Proxy WASM ABI Spec to integrate with Envoy and provide filtering and connectivity to Authorino and Limitador for request time enforcement of auth and rate limiting

Reviewer 5: Currently it's only Limitador. Authorino has no relation with the WASM Plugin yet.

### Dependencies and integrations

In order to provide its full suite of functionality, Kuadrant has several dependencies. Some of these are optional depending on the functionality needed.

- **[Cert-Manager](https://cert-manager.io/):** (provides the TLS integration)

>     - Optional: only required if you want to use TLSPolicy

Reviewer 5: Likely we'll have to add something about being required for auth, but only after https://github.com/Kuadrant/kuadrant-operator/issues/207. And until https://github.com/Kuadrant/architecture/pull/16 (or any other option to install Kuadrant without auth), we may need to flag cert-manager we required.

Gateway API provider that Kuadrant integrates with via WASM and Istio APIS to provide service protection capabilities.
- Required if you want to use RateLimitPolicy and AuthPolicy
- **[Gateway API](https://github.com/kubernetes-sigs/gateway-api):**
New standard for Ingress from the Kubernetes community.
- required. Core API we integrate with

## High Level Multi-Cluster Architecture

![](./images/high-level-multi-cluster.png)

Kuadrant has a multi-cluster gateway controller that is intended to run in a OCM provided "Hub" cluster. This cluster is effectively a central management cluster where policy and gateways along with all that OCM offers can be defined and distributed to the managed "spoke" clusters. 

## Single cluster 
In a single cluster context, the overall architecture remains the same as above, the key difference is that the Hub and Spoke cluster are now a single cluster rather than multiple clusters. This is how we are initially supporting single cluster.

## How does Kuadrant leverage OCM

The Kuadrant multi-cluster gateway controller is intended to be installed into an OCM hub cluster. The Gateway API resources are also installed into this cluster. 

> Kuadrant defines a [gateway class](https://gateway-api.sigs.k8s.io/api-types/gatewayclass/) that the multi-cluster gateway controller is the specified controller for. An OCM user can then (with the correct RBAC in place) define a Gateway just as they would in a single cluster setting and this will be reconciled by the multi-cluster gateway controller. We use this gateway definition to represent a "multi-cluster" enabled gateway and use the aggregated status from each instance of this gateway to store information such as IPs and Hostnames of Load Balancers and how many routes are attached to a given listner.

Reviewer 3: I got confused, too much too fast for me as the concepts flow from "class" to instances to and from the management cluster or something.

![](./images/ocm-integration.png)

### Gateway Deployment and Distribution

> In order for a multi-cluster gateway to be useful, it needs to be distributed or "placed" on a specific set of hub managed spoke clusters.  OCM is responsible for a set of placement and replication APIs. Kuadrant is aware of these APIs, and so when a given gateway is chosen to be placed on a set of managed clusters, Kuadrant multi-cluster gateway controller will ensure the right resources ([ManifestWork](https://open-cluster-management.io/concepts/manifestwork/)) are created in the correct namespaces in the hub. OCM then is responsible for syncing these to the actual spoke cluster and reporting back the status of these resources to the Hub. 

Reviewer 3: So what do I have to do as a user to tell I want this gateway in these two cluster?

In order for the Gateway to be instantiated, we need to know what gateway provider is being used on the spoke clusters. Admins can then set this provider in the hub via the Gateway class params. Kuadrant will then apply a transformation to ensure the gateway being synced references this spoke gateway provider (istio for example). 

> It is the OCM workagent that is responsible for syncing down and applying the resources into the managed spoke cluster. It is also responsible for syncing status information back to the hub. 

Reviewer 3: Some flow diagram with numbered callouts and description of the flow might be a useful alternative way to describe what is in this section.

The status information reported back to the Hub is used by the multi-cluster gateway controller to know what LB hosts / IPAddresses to use for DNSRecords that it creates and manages.

More info can be found here https://open-cluster-management.io/concepts/architecture/ 

## How does Kuadrant integrate with Gateway Providers

The Kuadrant data plane, integrates with the Istio gateway provider only currently: 

- It registers Authorino with the `IstioOperator` as an auth provider.

- It leverages an `EnvoyFilter`  to register the rate limiting service as a upstream service.

- It leverages istio `AuthorizationPolicy` to configure when Authorino should be called for a given host and path etc. 

> - It provides a WASM Plugin that conforms to the Proxy WASM ABI. This WASM Plugin is configured via the Kuadrant Operator based on valid rate limit policies and is then executed by the underlying Envoy instance that provides the proxy layer for the Gateway (IE Envoy). This plugin currently manages when RateLimiting and knows the location of the Limitador service.

Reviewer 3: "Manages when RateLimiting" and knows the location..."

I can't parse that sentence.

### Data Flows

There are several different data flows when using Kuadrant. 

**Control plane configuration and status reporting**

> The initial creation of these APIs (gateways, policies etc) is done by the relevant persona in the control plane just as they would any other k8s resource. We use the term cluster admin or gateway admin as the operatoins type persona configuring, and placing gateways. 

Reviewer 5: ```suggestion
The initial creation of these APIs (gateways, policies etc) is done by the relevant persona in the control plane just as they would any other k8s resource. We use the term cluster admin or gateway admin as the operations type persona configuring, and placing gateways. 
```

As shown above, in a multi-cluster configuration. API definitions are pulled from the Hub and "manifested" into the spokes. The Status of those synced resources are reported back to the Hub. The same happens for a single cluster, the only difference being the work agent hub controllers are all installed on one cluster.

**3rd party enforcement and Integration**

In order to enforce the policy configuration, components in the control plane and data plane can reach out to configured 3rd parties such as cloud based DNS provider, TLS providers and Auth providers.

**Request Flow**

> Requests coming through the gateway instance are intercepted by the WASM plugin installed into the Envoy based gateway provider. This plugin is then responsible for passing on the request to Limitador to be rate limited or Authorino to ensure auth. Each of these component each have the capability to see the request and need to in order to make the required decision. Each of these component can also prevent the request from reaching its backend destination based on user configuration. 

Reviewer 5: Same as https://github.com/Kuadrant/architecture/pull/24/files#r1309005383

![](./images/request-flow.png)

## Auth

> Authentication is handled by the underly kubernetes cluster. All APIs are CRDs and so authorization is controlled by RBAC in the same way as any kubernetes API.

Reviewer 5: Worth elaborating this a bit further perhaps? To make it clear it is about authenticating/authorising for using Kuadrant (as an App developer, as a cluster operator), and nothing to do with the "auth" Kuadrant delivers as a feature in the data plane. WDYT?

For Auth between Spoke and Hub see OCM docs https://open-cluster-management.io/concepts/architecture/

### Observability

Kuadrant doesn't provide any specific observability components, but rather provides a reference setup using well known and established components along with some useful dashboards to help observe key things around the Gateways. The focus of this setup, is in the context of a multi-cluster setup where OCM is installed and gateways are being defined and distributed from that hub.

![](./images/observability.png)

## Some notes on future direction

This section is here to provide some insight into architectural changes that may be seen in the near future:

What is in this doc represents the architecture at point our MVP release. Below are some areas that we have identified that are likely to change in the coming releases. As these happen, this doc will also evolve.

> - We want to separate out the ocm integration into its own controller

Reviewer 4: ```suggestion
- We want to separate out the ocm integration into its own controller so that policies can evolve without a coupling to any one multi-cluster management solution
```
I think this is the reason for this future direction?

As shown above, in a multi-cluster configuration. API definitions are pulled from the Hub and "manifested" into the spokes. The Status of those synced resources are reported back to the Hub. The same happens for a single cluster, the only difference being the work agent hub controllers are all installed on one cluster.

**3rd party enforcement and Integration**

In order to enforce the policy configuration, components in the control plane and data plane can reach out to configured 3rd parties such as cloud based DNS provider, TLS providers and Auth providers.

**Request Flow**

Requests coming through the gateway instance can be sent to Limitador based on configuration of the WASM plugin installed into the Envoy based gateway provider or to Authorino based configuration provided by the Istio `AuthorizationPolicy`. 
Each of these components have the capability to see the request and need to in order to make the required decision. Each of these components can also prevent the request from reaching its intended backend destination based on user configuration. 

> ![](./images/request-flow.png)

Reviewer 5: I don't think we need to change anything here necessarily, but I do wonder whether the depicted connection between Authorino and the external Auth Provider in the request flow diagram could be interpreted as a requirement, when actually it is not.

Partially, it may also be a matter of interpretation about what "external" means.

The part that is true is that Kuadrant does not manage user/client identities (i.e. Kuadrant is not an IAM system). In this sense, Authorino needs to go elsewhere to fetch something that will later allow it to verify the authentication tokens supplied in the requests.

That elsewhere can be an external auth provider (e.g. to fetch a set of JSON Web Keys) or the Kuberentes cluster itself (to pull API key secrets and x509 CA certs). Either way, definitely not part of the request flow, and perhaps not "external" depending on your POV.

Exceptional cases where Authorino will perform a request to a third-party authenticator every time (caching aside), as part of the request flow, are:
- Authentication based on OAuth opaque tokens;
- Kubernetes TokenReview – this one arguably not "external".

As I said, no need to change the diagram I think. It's good that integrating a proper auth provider is somehow represented. Devil's in the details tho. It's not always external, and it's rarely something that happens in the request flow.

Author: I think it is a fair point. I will merge as is for now so that the overview is present and visible but in future will make sure to take this into account.
